File: python/src/agent_framework/runtime/checkpointer_memory.py
"""In-memory checkpointer."""

from agent_framework.checkpoint import Checkpointer
from agent_framework.thread import ThreadStateSnapshot


class InMemoryCheckpointer(Checkpointer):
    """In-memory checkpointer for development / single-node."""

    def __init__(self) -> None:
        self._store: dict[str, list[ThreadStateSnapshot]] = {}

    async def put(self, snapshot: ThreadStateSnapshot) -> None:
        key = snapshot.thread_id
        if key not in self._store:
            self._store[key] = []
        self._store[key].append(snapshot)

    async def get(self, thread_id: str) -> ThreadStateSnapshot | None:
        list_snapshots = self._store.get(thread_id)
        if not list_snapshots:
            return None
        return list_snapshots[-1]

    async def list_snapshots(self, thread_id: str) -> list[ThreadStateSnapshot]:
        return self._store.get(thread_id, []).copy()

    async def delete(self, thread_id: str, node_id: str | None = None) -> None:
        if node_id is None:
            self._store.pop(thread_id, None)
            return
        list_snapshots = self._store.get(thread_id, [])
        self._store[thread_id] = [s for s in list_snapshots if s.node_id != node_id]
        if not self._store[thread_id]:
            del self._store[thread_id]

File: packages/runtime/src/checkpointer-memory.ts
import type { Checkpointer, ThreadStateSnapshot } from "@agent-framework/core";

/**
 * In-memory checkpointer for durable execution (development / single-node).
 */
export class InMemoryCheckpointer implements Checkpointer {
  private store = new Map<string, ThreadStateSnapshot[]>();

  async put(snapshot: ThreadStateSnapshot): Promise<void> {
    const key = snapshot.threadId;
    const list = this.store.get(key) ?? [];
    list.push(snapshot);
    this.store.set(key, list);
  }

  async get(threadId: string): Promise<ThreadStateSnapshot | null> {
    const list = this.store.get(threadId);
    if (!list?.length) return null;
    return list[list.length - 1] ?? null;
  }

  async list(threadId: string): Promise<ThreadStateSnapshot[]> {
    return this.store.get(threadId) ?? [];
  }

  async delete(threadId: string, nodeId?: string): Promise<void> {
    if (nodeId === undefined) {
      this.store.delete(threadId);
      return;
    }
    const list = this.store.get(threadId) ?? [];
    const next = list.filter((s) => s.nodeId !== nodeId);
    if (next.length) this.store.set(threadId, next);
    else this.store.delete(threadId);
  }
}
